### ISSUE TITLE
js file is not injected, but css file is injected to the template with 4.0.0-beta.11 version.

### ISSUE THREAD

**Issue by gracexiong** [jantimon/html-webpack-plugin#1355]
```js
new HtmlWebpackPlugin({
        filename: `${filePath + chunkName}.html`,
        template: `${ item }/${ chunkName }.html`,
        chunks: ['common'],
        cache: true,
        hash:true,
        inject: true
   });
```

**Comment by jantimon** [comment]
The filtering is happening here:

https://github.com/jantimon/html-webpack-plugin/blob/dfb98e70f813562c007249152ba902e53cbd94e1/index.js#L180

can you please add the following code just bellow that line to see what is happening?

```js
console.log({entryNames, chunks: self.options.chunks, filteredEntryNames});
```

**Comment by gracexiong** [comment]
{ entryNames:
   [ 'common/common',
     'es6Vendor',
     'app',
     'manager_Administrator_Modify',
     'manager_AdministratorReg',
     'registration',
     'account_banned',
     'base',
     'comment',
     'comment_center',
     'cps',
     'cps_apply',
     'data_analysis',
     'data_analysis_new',
     'demo',
     'drhomepage',
     'login',
     'manager/manager_bind',
     'manager/manager_cps',
     'manager/manager_data_analysis',
     'manager/manager_drhomepage',
     'manager/manager_index',
     'manager/manager_login',
     'manager/manager_post_mgmt',
     'msg_apply',
     'msg_center',
     'post_center',
     'post_center_old',
     'private_show',
     'registration_cpa',
     'registration_cpt',
     'tpl/tpl_art_center',
     'tpl/tpl_interlocution',
     'tpl/tpl_live',
     'vendor' ],
  chunks: [ 'common/common', 'private/private_show' ],
  filteredEntryNames: [ 'common/common' ] }

**Comment by jantimon** [comment]
But that seems to be the correct output right?

Especially this part:

```js
chunks: [ 'common/common', 'private_show' ],
filteredEntryNames: [ 'common/common', 'private_show' ] }
```

**Comment by gracexiong** [comment]
'entryPointFiles':[ 'css/private_show.css', 'js/private_show.js?4c019fd5' ]
 'assets' : { publicPath: '//xxx.com/',
  js: [],
  css: [ '/css/private_show.css?4c019fd5a9695d35bbd6' ],
  manifest: undefined,
  favicon: undefined }

**Comment by jantimon** [comment]
Okay so here it looks wrong - it should add the js file.

Even question marks should be supported:

https://github.com/jantimon/html-webpack-plugin/blob/dfb98e70f813562c007249152ba902e53cbd94e1/index.js#L598

Can you please also add a console.log here?

https://github.com/jantimon/html-webpack-plugin/blob/dfb98e70f813562c007249152ba902e53cbd94e1/index.js#L614

```
console.log({extensionRegexp, entryPointPublicPath})
```

**Comment by gracexiong** [comment]
{ extensionRegexp: /\.(css|js|mjs)(\?|$)/,
  entryPointPublicPath: '//xxx.com/css/private_show.css?4c019fd5a9695d35bbd6' }

**Comment by jantimon** [comment]
Does that mean the loop is executed only once? 🤔   
Do you have any guess where the data is disappearing?

**Comment by gracexiong** [comment]
I console in the loop
const extMatch = extensionRegexp.exec(entryPointPublicPath);
console.log(extMatch)

and  i got 

null
null
[ '.css?',
  'css',
  '?',
  index: 28,
  input: '//xxx.com/css/private_show.css?4c019fd5a9695d35bbd6',
  groups: undefined ]

the two null means these two js file not match?

**Comment by gracexiong** [comment]
i find it, because of add hash after js filename

/\.(css|js|mjs)(\?|$)/.exec('css/private_show.css')
(3) [".css", "css", "", index: 16, input: "css/private_show.css", groups: undefined]
/\.(css|js|mjs)(\?|$)/.exec('js/private_show.js?4c019fd5')
(3) [".js?", "js", "?", index: 15, input: "js/private_show.js?4c019fd5", groups: undefined]
/\.(css|js|mjs)(\?|$)/.exec('js/private_show.js')
(3) [".js", "js", "", index: 15, input: "js/private_show.js", groups: undefined]

**Comment by jantimon** [comment]
Sorry I can't follow - why would `const extensionRegexp = /\.(css|js|mjs)(\?|$)/; ` not work for ` 'js/private_show.js?4c019fd5'`?

**Comment by gracexiong** [comment]
/js/private_show.js%3F3252aa
the ? has been encodeURIComponent to %3F ,so didnot match? maybe you should decodeURIComponent  entry first?

**Comment by jantimon** [comment]
Okay that makes sense - where does the encoded `?` come from?

**Comment by gracexiong** [comment]
sorry for late reply, the "?"  is from "output:[name].js?[hash]" , i remove the "?[hash]" , and then it's ok, thank you so much

**Comment by jantimon** [comment]
I have reported this issue here: https://github.com/webpack/webpack/issues/10638

**Comment by jantimon** [comment]
It looks like the reason is this change by @zbigg who tried to allow whitespaces inside the url:

https://github.com/jantimon/html-webpack-plugin/commit/35a154186501fba3ecddb819b6f632556d37a58f

**Comment by zbigg** [comment]
@jantimon Well, my reasoning was that webpack works on "files", while `html-webpack-plugin` uses these files in HTML ans HTML expects `URL`s which shall be encoded.


TLDR That is not neccessarily true.

Now, looking at webpack documentation and this issue i see great impedance mismatch in meaning of "filename"
Some treat output as `URL`, some treat output as filename ([this doc doesn't refer to URL at all](https://webpack.js.org/configuration/output/#outputfilename).

So, looks like something that was expected to be filename was accidentally retrofitted to be "URL or filename"

> webpack is correctly removing the querystring before it writes files to disk however it is also encaping ? to %3F.

(https://github.com/webpack/webpack/issues/10638)

I would argue that it's not very correct, where you call something "filename" and then attempt to parse it as URL.

Nonetheless, giving that loads of people are relying on this behaviour, it's probably better to revert #1257 and document that "filename" is actually something between filename and URL and should be treated with care.
 


This should be resolved, i'm sorry i 
i see that `file name` is almost everywhere in documentation, but it's not super well specified whether "output" is filename (i.e. should be urlencoded before being "htmlized") or "expected" URL of resource (already urlencoded by user).

**Comment by imjeen** [comment]
I have a same problem with the latest version:  **html-webpack-plugin@4.0.3** .

when using the query output in webpack (a.k.a:  `output: { filename:  name].js?[contenthash]  } ` ),  the plugin didn't  include the JS files.

**Comment by jantimon** [comment]
@zbigg I am totally with you people are misusing the `output.filename` option (some add question marks others add plus signs).  

This has never been a supported or unit tested feature but a hack that worked because luckily webpack removes the querystring correctly before writing to disk.

But unfortunately people used this hack a lot and therefore we should probably allow to reuse it also for the new major release.

Does anyone know why `+` must not be escaped and is therefore excluded from `encodeURI`?

**Comment by jantimon** [comment]
@zbigg what do you think about #1386 ?

**Comment by jantimon** [comment]
Released as html-webpack-plugin 4.0.4
